File: app/services/query_types.py
"""查詢分派型別 — QueryKind 與 ParsedQuery 家族（#4 自 query_parse 抽出）。"""
from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Literal, Optional, Union

from app.services._generated.query_kind_registry import QueryKind

JYUTPING_ANCHOR_INVALID_HINT = (
    "粵拼錨無效：韻母片段喺收錄讀音中搵唔到對應。請檢查拼寫或改用漢字錨。"
)

# QueryKind SSOT: contracts/query-kind-manifest.json → codegen (ADR-0035)


@dataclass(frozen=True)
class RelationLookupQuery:
    relation_kind: Literal["syn", "ant"]
    word: str
    code_prefix: Optional[str] = None

    @property
    def kind(self) -> QueryKind:
        return QueryKind.RELATION_LOOKUP

    def to_handler_dict(self) -> dict:
        return {
            "kind": self.relation_kind,
            "word": self.word,
            "code_prefix": self.code_prefix,
        }


@dataclass(frozen=True)
class CompoundSynQuery:
    code_prefix: Optional[str]
    rhyme_char: Optional[str]

    @property
    def kind(self) -> QueryKind:
        return QueryKind.COMPOUND_SYN


@dataclass(frozen=True)
class CompoundDoubledSyllableQuery:
    width: int
    code_prefix: Optional[str]
    rhyme_char: Optional[str]

    @property
    def kind(self) -> QueryKind:
        return QueryKind.COMPOUND_DOUBLED_SYLLABLE


@dataclass(frozen=True)
class HeteronymCodeQuery:
    left_template: str
    right_template: str
    width: int

    @property
    def kind(self) -> QueryKind:
        return QueryKind.HETERONYM_CODE


@dataclass(frozen=True)
class CompoundAntQuery:
    code_prefix: Optional[str]
    rhyme_char: Optional[str]

    @property
    def kind(self) -> QueryKind:
        return QueryKind.COMPOUND_ANT


@dataclass(frozen=True)
class CompoundConnectSynQuery:
    code_prefix: Optional[str]
    connective: str
    rhyme_char: Optional[str]

    @property
    def kind(self) -> QueryKind:
        return QueryKind.COMPOUND_SYN


@dataclass(frozen=True)
class CompoundConnectAntQuery:
    code_prefix: Optional[str]
    connective: str
    rhyme_char: Optional[str]

    @property
    def kind(self) -> QueryKind:
        return QueryKind.COMPOUND_ANT


@dataclass(frozen=True)
class EqualsQuery:
    raw_q: str

    @property
    def kind(self) -> QueryKind:
        return QueryKind.EQUALS


@dataclass(frozen=True)
class PrefixWildcardEqualsQuery:
    raw_q: str
    inner_q: str
    ref_literal: str
    width: int

    @property
    def kind(self) -> QueryKind:
        return QueryKind.PREFIX_WILDCARD_EQUALS


@dataclass(frozen=True)
class PartialRhymeMaskQuery:
    raw_q: str
    pattern: str
    width: int
    anchors: list[tuple[int, str]]

    @property
    def kind(self) -> QueryKind:
        return QueryKind.PARTIAL_RHYME_MASK


@dataclass(frozen=True)
class PartialInitialMaskQuery:
    raw_q: str
    pattern: str
    width: int
    anchors: list[tuple[int, str]]

    @property
    def kind(self) -> QueryKind:
        return QueryKind.PARTIAL_INITIAL_MASK


@dataclass(frozen=True)
class SerialPhonemeAnchorQuery:
    raw_q: str
    width: int
    constraint: Literal["final", "initial"]
    code_slots: list[tuple[int, str]]
    anchors: list[tuple[int, str]]
    mask: str

    @property
    def kind(self) -> QueryKind:
        return QueryKind.SERIAL_PHONEME


@dataclass(frozen=True)
class PlusAnchorQuery:
    width: int
    constraint: str
    anchor: str
    anchor_pos: int
    code_slots: list[tuple[int, str]]
    code_prefix: Optional[str] = None

    @property
    def kind(self) -> QueryKind:
        return QueryKind.PLUS_ANCHOR

    def to_handler_dict(self) -> dict:
        return asdict(self)


@dataclass(frozen=True)
class WildcardCodeAnchorQuery:
    raw_q: str
    width: int
    slots: list[dict]
    head_literal: Optional[str] = None

    @property
    def kind(self) -> QueryKind:
        return QueryKind.WILDCARD_CODE_ANCHOR


@dataclass(frozen=True)
class CodeRefMiddleRhymeQuery:
    raw_q: str
    width: int
    anchor: str
    anchor_pos: int
    leading: str
    digits: str
    slots: list[dict]

    @property
    def kind(self) -> QueryKind:
        return QueryKind.CODE_REF_MIDDLE_RHYME


@dataclass(frozen=True)
class LiteralRefQuery:
    code_digits: str
    literal_char: str
    width: int

    @property
    def kind(self) -> QueryKind:
        return QueryKind.LITERAL_REF

    def to_handler_dict(self) -> dict:
        return asdict(self)


@dataclass(frozen=True)
class RhymeAnchorQuery:
    constraint: str
    anchor_pos: int
    anchor: str
    slots: str
    width: int

    @property
    def kind(self) -> QueryKind:
        return QueryKind.RHYME_ANCHOR

    def to_handler_dict(self) -> dict:
        return asdict(self)


@dataclass(frozen=True)
class JyutpingAnchorQuery:
    raw_q: str
    width: int
    anchor_pos: int
    anchor_kind: Literal["initial_letters", "rhyme_letters", "syllable_letters"]
    anchor_value: str
    code_prefix: Optional[str] = None
    code_slots: Optional[list] = None
    equals_style: bool = False
    hybrid_rhyme: bool = False
    dual_phoneme: bool = False
    dual_initial_value: Optional[str] = None

    @property
    def kind(self) -> QueryKind:
        return QueryKind.JYUTPING_ANCHOR


@dataclass(frozen=True)
class TripleRhymeAnchorQuery:
    anchor: str
    anchor_pos: int
    width: int
    leading_slots: str
    constraint: Literal["final"] = "final"

    @property
    def kind(self) -> QueryKind:
        return QueryKind.TRIPLE_RHYME_ANCHOR

    def to_handler_dict(self) -> dict:
        return asdict(self)


@dataclass(frozen=True)
class MaskQuery:
    raw_q: str

    @property
    def kind(self) -> QueryKind:
        return QueryKind.MASK


@dataclass(frozen=True)
class PingZeSerialQuery:
    raw_q: str
    pzmode: Literal["m1", "m2", "m3"] = "m1"
    anchor: Optional[str] = None

    @property
    def kind(self) -> QueryKind:
        return QueryKind.PING_ZE_SERIAL


@dataclass(frozen=True)
class DigitCodeQuery:
    raw_q: str

    @property
    def kind(self) -> QueryKind:
        return QueryKind.DIGIT_CODE


@dataclass(frozen=True)
class WordLookupQuery:
    raw_q: str

    @property
    def kind(self) -> QueryKind:
        return QueryKind.WORD_LOOKUP


@dataclass(frozen=True)
class JyutpingFragmentQuery:
    raw_q: str

    @property
    def kind(self) -> QueryKind:
        return QueryKind.JYUTPING_FRAGMENT


@dataclass(frozen=True)
class UnmatchedQuery:
    raw_q: str
    hint: Optional[str] = None

    @property
    def kind(self) -> QueryKind:
        return QueryKind.UNMATCHED


ParsedQuery = Union[
    RelationLookupQuery,
    CompoundSynQuery,
    CompoundDoubledSyllableQuery,
    HeteronymCodeQuery,
    CompoundAntQuery,
    CompoundConnectSynQuery,
    CompoundConnectAntQuery,
    EqualsQuery,
    PrefixWildcardEqualsQuery,
    PartialRhymeMaskQuery,
    PartialInitialMaskQuery,
    SerialPhonemeAnchorQuery,
    PlusAnchorQuery,
    WildcardCodeAnchorQuery,
    CodeRefMiddleRhymeQuery,
    LiteralRefQuery,
    RhymeAnchorQuery,
    TripleRhymeAnchorQuery,
    JyutpingAnchorQuery,
    MaskQuery,
    PingZeSerialQuery,
    DigitCodeQuery,
    WordLookupQuery,
    JyutpingFragmentQuery,
    UnmatchedQuery,
]

File: client/src/db/query-types.ts
import type { QueryKind } from './query-kind.ts';

export type QueryMode = 'm1' | 'm2' | 'm3' | '0243' | '02493' | '394052' | 'syn' | 'pz';

export interface ParsedQuery {
  kind: QueryKind;
  raw_q: string;
  hint?: string;
  pzmode?: 'm1' | 'm2' | 'm3';
  anchor?: string;
  base?: ParsedQuery;
  code_prefix?: string;
  relation_kind?: 'syn' | 'ant';
  connective?: string;
  rhyme_char?: string;
  word?: string;
}

export interface PingZeSerialQuery extends ParsedQuery {
  kind: QueryKind.PING_ZE_SERIAL;
  raw_q: string;
  pzmode: 'm1' | 'm2' | 'm3';
  anchor?: string;
  base?: ParsedQuery;
}

export interface DigitCodeQuery extends ParsedQuery {
  kind: QueryKind.DIGIT_CODE;
  raw_q: string;
}

export interface WordLookupQuery extends ParsedQuery {
  kind: QueryKind.WORD_LOOKUP;
  raw_q: string;
}

export interface JyutpingFragmentQuery extends ParsedQuery {
  kind: QueryKind.JYUTPING_FRAGMENT;
  raw_q: string;
}

export interface MaskQuery extends ParsedQuery {
  kind: QueryKind.MASK;
  raw_q: string;
}

export interface RhymeAnchorQuery extends ParsedQuery {
  kind: QueryKind.RHYME_ANCHOR;
  raw_q: string;
  constraint: 'final' | 'initial';
  anchor_pos: number;
  anchor: string;
  slots: string;
  width: number;
}

export interface PrefixWildcardEqualsQuery extends ParsedQuery {
  kind: QueryKind.PREFIX_WILDCARD_EQUALS;
  raw_q: string;
  inner_q: string;
  ref_literal: string;
  width: number;
}

export interface PartialRhymeMaskQuery extends ParsedQuery {
  kind: QueryKind.PARTIAL_RHYME_MASK;
  raw_q: string;
  pattern: string;
  width: number;
  anchors: Array<[number, string]>;
}

export interface PartialInitialMaskQuery extends ParsedQuery {
  kind: QueryKind.PARTIAL_INITIAL_MASK;
  raw_q: string;
  pattern: string;
  width: number;
  anchors: Array<[number, string]>;
}

export interface SerialPhonemeAnchorQuery extends ParsedQuery {
  kind: QueryKind.SERIAL_PHONEME;
  raw_q: string;
  width: number;
  constraint: 'final' | 'initial';
  code_slots: Array<[number, string]>;
  anchors: Array<[number, string]>;
  mask: string;
}

export interface JyutpingAnchorQuery extends ParsedQuery {
  kind: QueryKind.JYUTPING_ANCHOR;
  raw_q: string;
  width: number;
  anchor_pos: number;
  anchor_kind: 'initial_letters' | 'rhyme_letters' | 'syllable_letters';
  anchor_value: string;
  dual_phoneme?: boolean;
  code_prefix?: string;
  equals_style?: boolean;
  code_slots?: Array<[number, string]>;
  hybrid_rhyme?: boolean;
  dual_initial_value?: string;
}

export interface LiteralRefQuery extends ParsedQuery {
  kind: QueryKind.LITERAL_REF;
  raw_q: string;
  code_digits: string;
  literal_char: string;
  literal_pos: number;
  width: number;
}

export interface PlusAnchorQuery extends ParsedQuery {
  kind: QueryKind.PLUS_ANCHOR;
  raw_q: string;
  width: number;
  anchor_pos: number;
  anchor: string;
  constraint: 'literal' | 'final' | 'initial';
  code_slots: Array<[number, string]>;
  code_prefix?: string;
}

export interface CompoundSynQuery extends ParsedQuery {
  kind: QueryKind.COMPOUND_SYN;
  raw_q: string;
  code_prefix?: string;
  rhyme_char?: string;
}

export interface CompoundAntQuery extends ParsedQuery {
  kind: QueryKind.COMPOUND_ANT;
  raw_q: string;
  code_prefix?: string;
  rhyme_char?: string;
}

export interface CompoundConnectSynQuery extends ParsedQuery {
  kind: QueryKind.COMPOUND_CONNECT_SYN;
  raw_q: string;
  code_prefix?: string;
  connective: string;
  rhyme_char?: string;
}

export interface CompoundConnectAntQuery extends ParsedQuery {
  kind: QueryKind.COMPOUND_CONNECT_ANT;
  raw_q: string;
  code_prefix?: string;
  connective: string;
  rhyme_char?: string;
}

export interface CompoundDoubledSyllableQuery extends ParsedQuery {
  kind: QueryKind.COMPOUND_DOUBLED_SYLLABLE;
  raw_q: string;
  width: number;
  code_prefix?: string;
  rhyme_char?: string;
}

export interface HeteronymCodeQuery extends ParsedQuery {
  kind: QueryKind.HETERONYM_CODE;
  raw_q: string;
  left_template: string;
  right_template: string;
  width: number;
}

export interface WildcardCodeAnchorQuery extends ParsedQuery {
  kind: QueryKind.WILDCARD_CODE_ANCHOR;
  raw_q: string;
  width: number;
  slots: Array<{ pos: number; kind: string; value?: string }>;
  head_literal?: string;
}

export interface CodeRefMiddleRhymeQuery extends ParsedQuery {
  kind: QueryKind.CODE_REF_MIDDLE_RHYME;
  raw_q: string;
  width: number;
  anchor: string;
  anchor_pos: number;
  leading: string;
  digits: string;
  slots: Array<{ pos: number; kind: string; value?: string }>;
}

export interface TripleRhymeAnchorQuery extends ParsedQuery {
  kind: QueryKind.TRIPLE_RHYME_ANCHOR;
  raw_q: string;
  anchor: string;
  anchor_pos: number;
  width: number;
  leading_slots: string;
  constraint: 'final';
}

export interface RelationLookupQuery extends ParsedQuery {
  kind: QueryKind.RELATION_LOOKUP;
  raw_q: string;
  relation_kind: 'syn' | 'ant';
  word: string;
  code_prefix?: string;
}

export interface UnmatchedQuery extends ParsedQuery {
  kind: QueryKind.UNMATCHED;
  raw_q: string;
  hint?: string;
}

export interface QueryResult {
  word: string;
  jyutping: string;
  code: string;
  definition?: string;
  score: number;
  char?: string;
  display_text?: string;
  /** ponytail: lookup layout row kind — upgrade path: full lookup_layout.ts module */
  resultType?: 'code' | 'jyutping' | 'word';
  heteronym_tags?: string[];
  anchor_dimension?: 'initial' | 'final';
  relation?: 'syn' | 'ant' | 'semantic_related';
  in_db?: boolean;
  source?: string;
}

export interface SearchContext {
  q: string | null;
  code?: string;
  char?: string;
  mode: QueryMode;
  pzmode?: 'm1' | 'm2' | 'm3';
  limit: number;
  offset: number;
  fallback_0243_mode?: QueryMode;
  ui_lang?: 'zh' | 'zh-Hans' | 'en';
  /** Cooperative cancel — engine hot paths throw SearchCancelledError when true */
  shouldCancel?: () => boolean;
}

export interface SearchResult {
  items: QueryResult[];
  total?: number;
  hint?: string;
  cache_path?: string;
  effective_mode?: QueryMode;
  /** 純漢字詞條 lookup：只輸出詞列（PWA 詞條行已含碼／粵拼；見 CONTEXT 詞條 lookup 版面） */
  lookup_layout?: boolean;
}
